### PR TITLE
Dialogue choices, conditions added. Graph traversal refactored

### PR DESCRIPTION
**Pull request by BDeshiDev**
@coderabbitai review

## Summary by CodeRabbit

- **New Features**
	- Automated workflow for updating and publishing UPM branches.
	- Automated documentation generation and deployment workflow.

- **Documentation**
	- Updated CHANGELOG.md with recent changes.
	- Added licensing information in LICENSE.md.

- **Enhancements**
	- Introduced new classes and prefabs for dialogue system enhancements.
	- Added new UI components for dialogue choices.
	- Implemented node-based dialogue graph management.
	- Improved character data models using Scriptable Objects.
	- Custom node editor for dialogue system.
	- New shaders and materials for TextMesh Pro assets optimized for mobile platforms.

- **Refactor**
	- Codebase refactoring for a better public API.
	- Refactored event handling and UI code for decoupling.

- **Chores**
	- Updated .gitignore for Unity-specific files.
	- Added editor configurations for Rider IDE.

- **Style**
	- New font materials and style sheets for TextMesh Pro.

- **Project Settings**
	- Updated project settings for audio, input, graphics, physics, editor, and package management.

### CODE REVIEW

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/IDialogueNodeCondition.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+namespace Packages.com.studio23.ss2.dialoguesystem.Runtime.Data
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Data/IDialogueNodeCondition.cs]: Fix namespace

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueSystem.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+        /// <summary>
+        /// Subscribe to this to know when dialogue completed
+        /// </summary>
+        public DialogueLineEvent DialogueLineCompleted;
+        
+
+        void Awake()
+        {
+            Instance = this;
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueSystem.cs:39]: Not a good idea to use lazy singleton instantiation

File: .github/workflows/UPMBranchUpdate.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+name: Update and Publish UPM Branch
```
Comment (Warhammer4000) [inline .github/workflows/UPMBranchUpdate.yml:1]: This needs to be updated from the Loop

File: Assets/Packages/com.studio23.ss2.dialoguesystem/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Changelog
+
```
Comment (Warhammer4000) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/CHANGELOG.md:6]: you should provide a changelog as major changes happened

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/Configuration/SubtitleSettings.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/Configuration/SubtitleSettings.cs]: Empty space.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/Configuration/SubtitleSettings.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+[CreateAssetMenu(fileName = "SubtitleSettings", menuName = "Studio-23/Dialogue System/Subtitle Settings")]
+public class SubtitleSettings : ScriptableObject
+{
+
+    public Color SubtitleColor = Color.black;
+    public int SubtitleFontSize = 24;
+    public bool DropShadow = false;
+
+
```
Comment (N-Evan) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/Configuration/SubtitleSettings.cs]: Add option to change font from editor?

**Review by N-Evan — APPROVED** [review]

**Review by N-Evan — APPROVED** [review]

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueSystem.cs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            _currentGraph = newGraph;
+        }
+
+
+        /// <summary>
+        /// Start playing dialogue and it Fires the Start event
+        /// </summary>
+        public void StartDialogue()
+        {
+            PlayDialogue();
```
Comment (gitOpu) [inline Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueSystem.cs:57]: StartDialogue method calls an asynchronous method PlayDialogue

**Review by gitOpu — APPROVED** [review]
everything looks good.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueSystem.cs (hunk 7 above)
Comment (BDeshiDev) [reply Assets/Packages/com.studio23.ss2.dialoguesystem/Runtime/Core/DialogueSystem.cs:57]: It's there if you just want to start the dialogue but don't want to wait.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/DialogueBoxUI.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    
+
+    private void RegisterEvents()
+    {
+        DialogueSystem.Instance.OnDialogueStarted +=  ShowUI;
+        DialogueSystem.Instance.OnDialogueEnded +=  HideUI;
+        DialogueSystem.Instance.DialogueLineStarted += handleDialogueLineStarted;
+
+        DialogueSystem.Instance.OnDialogueChoiceStarted += HideUI;
+        DialogueSystem.Instance.OnDialogueChoiceEnded += ShowUI;
```
[the remark this answers is not in the document]
Comment (BDeshiDev) [reply Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/DialogueBoxUI.cs:56]: We assume that the singleton will always be there. When OnDestroy is called after the game closes, the singleton might have been destroyed when we try to unsub. That's why an explicit null check is needed

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/TextMeshProLocalizer.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    /// TMP immediately shows tempText.
+    /// Localization load done return actual string in correct language
+    /// </summary>
+    /// <param name="tempText"></param>
+    /// <param name="localizedString"></param>
+    public async UniTask<string> LoadTextAndWait(LocalizedString localizedString, string tempText= "")
+    {
+        SetText(localizedString, tempText);
+        return await _localizer.StringReference.GetLocalizedStringAsync();
+    }
```
[the remark this answers is not in the document]
Comment (BDeshiDev) [reply Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/TextMeshProLocalizer.cs:79]: This is a sample UI.

File: Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/DialogueBoxUI.cs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            DialogueSystem.Instance.OnDialogueChoiceStarted -= HideUI;
+            DialogueSystem.Instance.OnDialogueChoiceEnded -= ShowUI;
+        }
+    }
+
+ 
+    private void handleDialogueLineStarted(DialogueLineNodeBase dialogueLineNodeBase)
+    {
+        ShowDialogueTextAsync(dialogueLineNodeBase);
+    }
```
[the remark this answers is not in the document]
Comment (BDeshiDev) [reply Assets/Packages/com.studio23.ss2.dialoguesystem/Samples/DialogueChoiceDemo/DialogueBoxUI.cs:76]: This is needed to match function signatures. We don't care about waiting for the UI text to fully load.

**Pull request merged**